src/library/masq.js

```diff
@@ -450,72 +515,81 @@ class Masq {
-      }
+    await this.sendAccessGranted(this.peer)
+  }
 
-      const sendAccessGranted = async (peer, dbKey, userAppDbId, userAppDEK, username, profileImage, userAppNonce) => {
-        const data = { msg: 'masqAccessGranted', key: dbKey, userAppDbId, userAppDEK, username, profileImage, userAppNonce }
-        const encryptedMsg = await encrypt(this.key, data, 'base64')
-        peer.send(JSON.stringify(encryptedMsg))
-      }
+  async sendAccessRefused (peer) {
+    this.setState(STATES.USER_REFUSED)
+    const data = { msg: 'masqAccessRefused' }
+    const encryptedMsg = await encrypt(this.key, data, 'base64')
+    peer.send(JSON.stringify(encryptedMsg))
+  }
 
-      const sendWriteAccessGranted = async (peer) => {
-        const data = { msg: 'writeAccessGranted' }
-        const encryptedMsg = await encrypt(this.key, data, 'base64')
-        peer.send(JSON.stringify(encryptedMsg))
-      }
+  async sendAccessGranted (peer) {
+    this.setState(STATES.USER_ACCEPTED)
+    const apps = await this.getApps()
+    const app = apps.find(app => app.appId === this.appId)
+    const appDEK = app ? app.appDEK : await this._genAppDEK()
+    const appNonce = app ? app.appNonce : await this._genNonce()
 
-      const handleData = async (peer, data) => {
-        const json = await decrypt(this.key, JSON.parse(data), 'base64')
-        const { msg } = json
-        // TODO: Error if  missing params
-
-        if (msg === 'requestWriteAccess') {
-          const userAppKey = Buffer.from(json.key, 'hex')
-          try {
-            await this.appsDBs[dbName].authorizeAsync(userAppKey)
-          } catch (err) {
-            throw new MasqError(MasqError.AUTHORIZE_DB_KEY_FAILED)
-          }
-          await sendWriteAccessGranted(peer)
-          this.sw.close()
-          return resolve()
-        }
-      }
+    const id = app ? app.id : await this.addApp({
+      ...this.app, appId: this.appId, appDEK, appNonce
+    })
 
-      if (!isGranted) {
-        await sendAccessRefused(this.peer)
-        await this._closeUserAppConnection()
-        return resolve()
+    const privateProfile = await this.getProfile()
+
+    const dbName = `app-${this.profileId}-${id}`
+    const db = await this._createDBAndSyncApp(dbName)
+    // this.dbName must be filled only during the userApp register protocol
+    this.dbName = dbName
+
+    const dbKey = db.key.toString('hex')
+    const userAppDbId = id
+    const userAppDEK = appDEK
+    const username = privateProfile.username
+    const profileImage = privateProfile.image
+    const userAppNonce = appNonce
+
+    const data = { msg: 'masqAccessGranted', key: dbKey, userAppDbId, userAppDEK, username, profileImage, userAppNonce }
+    const encryptedMsg = await encrypt(this.key, data, 'base64')
+    peer.send(JSON.stringify(encryptedMsg))
+    peer.once('data', async (data) => {
+      const json = await decrypt(this.key, JSON.parse(data), 'base64')
+
+      switch (json.msg) {
```

Comment:
If we have only one type of msg, I would prefer we use an if statement instead of switch/case

Reply:
yes I add the if statement and separate the on('data') into a separate method.